contracts/EventBasedPredictionMarket.sol

```diff
@@ -0,0 +1,325 @@
+     * @param timestamp The timestamp of the price request.
+     * @param ancillaryData Custom ancillary data to be passed along with the price request to the OO.
+     * @param refund The amount of collateral refunded to the caller of the price request.
+     */
+    function priceDisputed(
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData,
+        uint256 refund
+    ) external {
+        OptimisticOracleInterface optimisticOracle = _getOO();
+        require(msg.sender == address(optimisticOracle), "not authorized");
+
+        expirationTimestamp = getCurrentTime();
+        require(timestamp <= expirationTimestamp, "different timestamps");
+        require(identifier == priceIdentifier, "same identifier");
+        require(keccak256(ancillaryData) == keccak256(customAncillaryData), "same ancillary data");
+        require(refund == proposerReward, "same proposerReward amount");
+
+        _requestOraclePrice();
+    }
+
+    /****************************************
+     *          POSITION FUNCTIONS          *
+     ****************************************/
+
+    /**
+     * @notice Creates a pair of long and short tokens equal in number to tokensToCreate. Pulls the required collateral
+     * amount into this contract, defined by the collateralPerPair value.
+     * @param tokensToCreate number of long and short synthetic tokens to create.
+     * @return collateralUsed total collateral used to mint the synthetics.
+     */
+    function create(uint256 tokensToCreate) public requestInitialized returns (uint256 collateralUsed) {
+        // Note the use of mulCeil to prevent small collateralPerPair causing rounding of collateralUsed to 0 enabling
+        // callers to mint dust LSP tokens without paying any collateral.
+        collateralUsed = FixedPoint.Unsigned(tokensToCreate).mulCeil(FixedPoint.Unsigned(collateralPerPair)).rawValue;
+
+        collateralToken.safeTransferFrom(msg.sender, address(this), collateralUsed);
+
+        require(longToken.mint(msg.sender, tokensToCreate));
+        require(shortToken.mint(msg.sender, tokensToCreate));
+
+        emit TokensCreated(msg.sender, collateralUsed, tokensToCreate);
+    }
+
+    /**
+     * @notice Redeems a pair of long and short tokens equal in number to tokensToRedeem. Returns the commensurate
+     * amount of collateral to the caller for the pair of tokens, defined by the collateralPerPair value.
+     * @param tokensToRedeem number of long and short synthetic tokens to redeem.
+     * @return collateralReturned total collateral returned in exchange for the pair of synthetics.
+     */
+    function redeem(uint256 tokensToRedeem) public returns (uint256 collateralReturned) {
+        require(longToken.burnFrom(msg.sender, tokensToRedeem));
+        require(shortToken.burnFrom(msg.sender, tokensToRedeem));
+
+        collateralReturned = FixedPoint.Unsigned(tokensToRedeem).mul(FixedPoint.Unsigned(collateralPerPair)).rawValue;
```

Comment:
my onle remaining feedback is we might want to remove this fixedPoint math throughout this contract?